### PR TITLE
A Dart implementation

### PR DESCRIPTION
**Pull request by Hixie**
Not sure if you are interested in having other implementations, but if you are, here is one. :-)

### CODE REVIEW

File: src/dart/lib/decoder.dart
Comment (dnfield) [inline src/dart/lib/decoder.dart:66]: Seems strange to have a public undocumented member for tests in this file.

File: src/dart/lib/decoder.dart
Comment (dnfield) [inline src/dart/lib/decoder.dart:235]: Consider using dart style on these (e.g. `creg`, `nreg`, `h`, but spelled out fully without abbreviations).

File: src/dart/lib/decoder.dart
Comment (dnfield) [inline src/dart/lib/decoder.dart:269]: nit: meaningful string in state error message

File: src/dart/lib/decoder.dart
Comment (dnfield) [inline src/dart/lib/decoder.dart]: nit: I think the style usually used for this is `onDebugDescribe?.call(....)`

File: src/dart/lib/decoder.dart
Comment (dnfield) [inline src/dart/lib/decoder.dart]: It's weirder that Flutter doesn't use premul alpha here - and if we ever "fix" that we should update this (https://github.com/flutter/flutter/issues/48674).

File: src/dart/lib/decoder.dart
Comment (dnfield) [inline src/dart/lib/decoder.dart:174]: this method probably shouldn't be public right? I understand this is a private class, but calling this a second time after construction will always be an error.

It might make more sense to just have a `computePicture` method or something like that instead of `execute`, and get rid of the late member field. I'd typically expect a ctor to be cheap, so that I can instantiate the machine wherever I want and control when/how often it actually does work. It may also be good to make that an async function, since you may at some point want it to do some of its work async in a separate isolate or something like that. Obviously can be avoided if you're super confident there will never be a good reason to have async parsing (for example, it has to interact with dart:ui objects too frequently to allow for any benefit from async computation/io).

File: src/dart/lib/widgets.dart
Comment (dnfield) [inline src/dart/lib/widgets.dart:51]: Is there any way we can pass this through the palette? I'm assuming no because we might have to blend things in some complicated manner (e.g. draw paths of multiple alpha values over each other and then alpha blend the whole thing at the end). But this will be expensive :\

File: src/dart/lib/widgets.dart
Comment (dnfield) [inline src/dart/lib/widgets.dart]: I'm pretty sure this will not perform very well. It will be doing decodes on every build.

In flutter_svg, I ended up creating a `RawPicture` widget patterned after `RawImage` and a `PictureCache` to help with this. I wonder if we should just create something like that at the framework level at this point.

File: src/dart/lib/widgets.dart
Comment (dnfield) [inline src/dart/lib/widgets.dart]: It would probably help to see the suggestion on changes to `execute` above bubble up here. This call looks cheap (just constructing an object), can actually be quite expensive and probably should never be used in a synchronous build method.

File: src/dart/lib/widgets.dart
Comment (dnfield) [inline src/dart/lib/widgets.dart]: It would be good to avoid using a LayoutBuilder here - maybe just some way to quickly parse the size of the IconVG before/while parsing the rest of the file would work better.

File: src/dart/lib/widgets.dart
Comment (dnfield) [inline src/dart/lib/widgets.dart:160]: Consider adding a `semanticsLabel` and bubbling that up/down.

File: src/dart/lib/decoder.dart
Comment (dnfield) [reply src/dart/lib/decoder.dart:66]: Consider marking this `@visibleForTesting` and documenting where it's used. Alternatively, consider making the implementation you're testing public or `@visibleForTesting` so that we don't have test code living in the same file as the impl.

**Review by dnfield — COMMENTED** [review]
Neat!

The PicturePainter classes are very similar to some widgets in flutter_svg. I wonder if we want to move some of that logic to either the framework or a separate package that both of these depend on. IME, it's not a super complicated widget to start with but as you want to do things like caching it gets more complicated.

File: src/dart/lib/decoder.dart
Comment (Hixie) [reply src/dart/lib/decoder.dart:235]: I went with matching the spec for sanity.

File: src/dart/lib/decoder.dart
Comment (Hixie) [reply src/dart/lib/decoder.dart:269]: Not really sure what to put here. "Compiler should be able to detect that this is unreachable"? :-)

File: src/dart/lib/decoder.dart
Comment (Hixie) [reply src/dart/lib/decoder.dart]: oops, meant to remove all this debug logic.

(FWIW I personally try to avoid `call` as much as possible.)

File: src/dart/lib/decoder.dart
Comment (Hixie) [reply src/dart/lib/decoder.dart]: While implementing this I was thinking the opposite -- that I wish this used straight alpha. :-)

Premultiplied alpha is less expressive.

File: src/dart/lib/decoder.dart
Comment (Hixie) [reply src/dart/lib/decoder.dart:174]: `execute` isn't really special here, the `checkSignature` and `readMetadata` methods are equally one-time-only. The whole class is really just a very large function with nested functions. I guess I could switch to that style instead, but I liked how the class syntax guarantees there's no stray imperative code after the constructor's `}` that might do something.

File: src/dart/lib/decoder.dart
Comment (dnfield) [reply src/dart/lib/decoder.dart:269]: I was thinking just something like `Unsupported tile mode ${(color & 0x00C00000).toRadixString(16)}`

File: src/dart/lib/decoder.dart
Comment (dnfield) [inline src/dart/lib/decoder.dart:571]: (same nit as above for 'unreachable')

File: src/dart/lib/widgets.dart
Comment (Hixie) [reply src/dart/lib/widgets.dart:51]: Yeah.

File: src/dart/lib/widgets.dart
Comment (Hixie) [reply src/dart/lib/widgets.dart]: Which part in particular do you think will be expensive?

File: src/dart/lib/widgets.dart
Comment (Hixie) [reply src/dart/lib/widgets.dart]: IconVG files are quite small and really fast to parse. On my workstation (in JIT mode, as part of `flutter test`), most of the test images are parsed in less than 1ms, on my Pixel 3XL in release mode, they parse in a few hundred microseconds (the slowest test image seems to be `video-005.primitive.ivg` at 236µs). Doing this asynchronously only really makes sense if the work could be spread over multiple frames, but making this interruptible seems like it would add more overhead than just the cost of rendering in the first place.

It's possible that more complicated images would take longer to render, of course, but until this format gets used, it seems like a premature optimization to worry about that.

File: src/dart/lib/widgets.dart
Comment (Hixie) [reply src/dart/lib/widgets.dart]: The actual rendered height is what is being obtained here. IconVGs have a level-of-detail feature that allows them to render differently based on the rasterization size.

File: src/dart/lib/widgets.dart
Comment (Hixie) [reply src/dart/lib/widgets.dart:160]: Done.

File: src/dart/lib/decoder.dart
Comment (Hixie) [reply src/dart/lib/decoder.dart:66]: `@visibleForTesting` is just another way of writing `public`. :-)

I figured this will all get tree-shaken so it doesn't really matter. Added docs though.

File: src/dart/lib/decoder.dart
Comment (Hixie) [reply src/dart/lib/decoder.dart:269]: The only way this could get hit is if I made a typo somewhere, and if I did, then it's very likely the message would be wrong...

File: src/dart/lib/widgets.dart
Comment (dnfield) [reply src/dart/lib/widgets.dart]: This will involve re-parsing the bytes every time the widget is built, right?

File: src/dart/lib/widgets.dart
Comment (dnfield) [reply src/dart/lib/widgets.dart]: A pixel 3 is still significantly more powerful than the average Android device worldwide. That said, these times aren't particularly concerning, but it'd be worth testing something that e.g. draws a gradient across the entirety of the screen.

It'd also be worth looking at what the CPU utilization is - even if it's fast, it's still using more cycles than if you cached this, especially since you know it won't change from frame to frame.

File: src/dart/lib/widgets.dart
Comment (dnfield) [reply src/dart/lib/widgets.dart]: Yes, but it's being obtained after doing a layout pass because of the layout builder - or am I missing something?

**Comment by nigeltao** [comment]
Wow!

I'm happy to merge the PR whenever @dnfield is happy from a Dart/Flutter expertise code review perspective.

File: src/dart/lib/decoder.dart
Comment (nigeltao) [reply src/dart/lib/decoder.dart]: FWIW, in what I'd expect to be a 'fade out' common case, a gradient from `(R, G, B, A)` to `(R, G, B, 0)` in nonpremul space is equivalent to a gradient from `(R, G, B, A)` to `(0, 0, 0, 0)` in premul space. (I think; you might need to check my math).

For lerping from opaque pink to zero-alpha orange (not zero-alpha pink) then, yes, premul alpha is less expressive. But you might be able to approximate it by inserting multiple intermediate stops, including an ε-alpha orange just before the `(0, 0, 0, 0)` stop...

**Comment by nigeltao** [comment]
Although, as you're aware, the file format will probably change soon (https://github.com/google/iconvg/issues/4). I'll leave it up to you whether you want this PR to land before or after that happens.

**Comment by Hixie** [comment]
If you don't mind some lag between the spec changing and fixes landing here, I'll land this without waiting for the spec updates.

File: src/dart/lib/widgets.dart
Comment (Hixie) [reply src/dart/lib/widgets.dart]: Not sure I follow. LayoutBuilder builds during layout, it's still single-pass.

Comment (dnfield) [reply src/dart/lib/widgets.dart]: Ok - this should be fine then.

File: src/dart/lib/widgets.dart
Comment (Hixie) [reply src/dart/lib/widgets.dart]: Not sure I follow. We are caching the Picture from frame to frame here. What do you want to cache?

I can profile this to see where the expense in the parsing lies. I've done no work to optimize this so far. It's quite possible there's low-hanging fruit.

File: src/dart/lib/widgets.dart
Comment (Hixie) [reply src/dart/lib/widgets.dart]: Only if the bytes, palette, size, or clip behaviour changes.

File: src/dart/lib/widgets.dart
Comment (dnfield) [inline src/dart/lib/widgets.dart:221]: May not be necessary for this PR, but I wonder how this will do if you have a `ListView` with lots of entries and the same `IconVG` in every entry.

That could probably be easily solved by having the `IconVG` widget take an `IconVGFile` as a parameter, and make the .asset ctor a convenience constructor that loads that from an asset repeatedly. Caching could also always be added later in a non-breaking way though.

**Review by dnfield — APPROVED** [review]
LGTM, anything else here are just nitpicks.

File: src/dart/lib/widgets.dart
Comment (Hixie) [reply src/dart/lib/widgets.dart:221]: Ah yeah, interesting. Ideally we'd be able to plug into the ImageCache logic but that would require making that work with some abstract superclass of Image and Picture...

File: src/dart/lib/widgets.dart
Comment (Hixie) [reply src/dart/lib/widgets.dart]: (unless i got the logic wrong)

Comment (dnfield) [reply src/dart/lib/widgets.dart]: Yeah, I missed something here, this should be fine.

File: src/dart/lib/widgets.dart
Comment (dnfield) [reply src/dart/lib/widgets.dart:221]: Pictures are harder because there's no great way to tell their actual total size, particularly if they have `Image` objects embedded within them (wouldn't happen with this implementation but can with others, e.g. svg).

**Comment by Hixie** [comment]
@nigeltao Sounds like we're ready to merge if you want to merge it. (@dnfield let me know if I missed anything.)

**Pull request merged**
